src-control/Control/RefCount.hs

```diff
@@ -131,3 +165,231 @@ readRefCount RefCounter{countVar} = RefCount <$> readPrimVar countVar
+#endif
+
+#ifdef NO_IGNORE_ASSERTS
+assertNotReleased :: PrimMonad m => IORef Bool -> m ()
+assertNotReleased released = do
+    r <- unsafeIOToPrim $ readIORef released
+    assert (not r) $ return ()
+#endif
+
+-- | Class of objects which support 'Ref'.
+--
+-- For objects in this class but not 'HasSharedFinaliser', the guarantee is
+-- that (when the 'Ref' rules are followed) the 'unsharedFinaliser' is called
+-- exactly once.
+--
+-- For objects that are also in 'HasSharedFinaliser', the 'unsharedFinaliser'
+-- is repurposed for reference counting, but the guarantee instead is that the
+-- object's 'sharedFinaliser' will be called exactly once.
+--
+class HasFinaliser obj where
+  type FinaliserM obj :: Type -> Type
+  unsharedFinaliser :: FinaliserM obj ~ m => obj -> m ()
+
+  default unsharedFinaliser :: (HasSharedFinaliser obj, FinaliserM obj ~ m,
+                                PrimMonad m, MonadMask m)
+                            => obj -> m ()
+  unsharedFinaliser obj =
+    mask_ $ do
+      done <- decrementRefCounter (getRefCounter obj)
+      when done (sharedFinaliser obj)
+
+-- | Class of objects that support 'Ref' including 'dupRef', with the guarantee
+-- that the object's 'sharedFinaliser' will be called exactly once.
+--
+-- Do not define 'unsharedFinaliser' for types in this class. The default
+-- implementation is used for the reference counting.
+--
+class HasSharedFinaliser obj where
+  getRefCounter   :: obj -> RawRefCounter (FinaliserM obj)
+
+  sharedFinaliser :: FinaliserM obj ~ m => obj -> m ()
+
+newRef :: PrimMonad m => obj -> m (Ref obj)
+#ifndef NO_IGNORE_ASSERTS
+newRef obj = return (Ref obj)
+#else
+newRef obj = do
+    released <- unsafeIOToPrim $ newIORef False
+    return (Ref obj released)
+#endif
+
+-- | Convenience helper to make an object containing a reference count.
+newSharedRef :: (HasSharedFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+             => (RawRefCounter m -> obj) -> m (Ref obj)
+newSharedRef mkObject = do
+    rc  <- newRawRefCounter
+    let !obj = mkObject rc
+    assert (getRefCounter obj == rc) $
+      newRef obj
+
+-- | Release a reference to an object that will no longer be used (via this
+-- reference).
+--
+releaseRef :: (HasFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+           => Ref obj -> m ()
+#ifndef NO_IGNORE_ASSERTS
+releaseRef (Ref obj) = unsharedFinaliser obj
+#else
+releaseRef (Ref obj released) = do
+    assertNotReleased released
+    unsharedFinaliser obj
+    unsafeIOToPrim $ writeIORef released True
+#endif
+
+-- | Use the object in a 'Ref'. Do not retain the object after the scope of
+-- the body. If you cannot use scoped \"with\" style, use 'unsafeDeRef'.
+--
+#ifndef NO_IGNORE_ASSERTS
+withRef :: Ref obj -> (obj -> m a) -> m a
+withRef (Ref obj) f = f obj
+#else
+withRef :: PrimMonad m => Ref obj -> (obj -> m a) -> m a
+withRef (Ref obj released) f = assertNotReleased released
+                            >> f obj
+#endif
```

Comment:
It might be nicer to have the same type signature in both cases, so we can't run into unused constraint warnings at the usage sites that only show up in non-debug builds. This means we'd have to deal with the redundant constraint issue here, but it's probably the right place.

Reply:
Yes. Done. I've kept the HasCallstack conditional though, since that has a runtime cost.